python/lsst/pipe/base/_dataset_handle.py

```diff
@@ -0,0 +1,153 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ["InMemoryDatasetHandle"]
+
+import dataclasses
+from typing import Any, Optional
+
+from lsst.daf.butler import DataCoordinate, DimensionUniverse, StorageClass, StorageClassFactory
+from lsst.utils.introspection import get_full_type_name
+
+# Use an empty dataID as a default.
+_DEFAULT_DATAID = DataCoordinate.standardize({}, universe=DimensionUniverse())
+
+
+@dataclasses.dataclass(frozen=True)
+class InMemoryDatasetHandle:
+    """An in-memory version of a `~lsst.daf.butler.DeferredDatasetHandle`."""
+
+    def get(
+        self, *, component: Optional[str] = None, parameters: Optional[dict] = None, **kwargs: dict
+    ) -> Any:
+        """Retrieves the dataset pointed to by this handle
+
+        This handle may be used multiple times, possibly with different
+        parameters.
+
+        Parameters
+        ----------
+        component : `str` or None
+            If the deferred object is a component dataset type, this parameter
+            may specify the name of the component to use in the get operation.
```

Comment:
Did you mean to write "if the deferred object is a composite dataset type"?

Does this all only supports the case where you can initialize the handle with the full composite, not just the component that you'll later need?

Reply:
If someone gives this a component I think that's no different to it being given any other type of non-composite object. People don't do `.get(component="x")` if they had already passed in component `x` do they? I might be missing something.

Reply:
No, I think you're right - if someone starts with a component, then it just works.